pkg/monitoring/vms/prometheus/prometheus.go

```diff
@@ -77,385 +77,180 @@ func tryToPushMetric(desc *prometheus.Desc, mv prometheus.Metric, err error, ch
-		)
-		tryToPushMetric(memoryAvailableDesc, mv, err, metrics.ch)
 	}
 
-	if vmStats.Memory.UnusedSet {
-		memoryUnusedLabels := []string{"node", "namespace", "name"}
-		memoryUnusedLabels = append(memoryUnusedLabels, metrics.k8sLabels...)
-		memoryUnusedDesc := prometheus.NewDesc(
+	if mem.UnusedSet {
+		metrics.pushCommonMetric(
 			"kubevirt_vmi_memory_unused_bytes",
 			"amount of unused memory as seen by the domain.",
-			memoryUnusedLabels,
-			nil,
-		)
-
-		memoryUnusedLabelValues := []string{metrics.vmi.Status.NodeName, metrics.vmi.Namespace, metrics.vmi.Name}
-		memoryUnusedLabelValues = append(memoryUnusedLabelValues, metrics.k8sLabelValues...)
-		mv, err := prometheus.NewConstMetric(
-			memoryUnusedDesc, prometheus.GaugeValue,
-			// the libvirt value is in KiB
-			float64(vmStats.Memory.Unused)*1024,
-			memoryUnusedLabelValues...,
+			prometheus.GaugeValue,
+			float64(mem.Unused)*1024,
 		)
-		tryToPushMetric(memoryUnusedDesc, mv, err, metrics.ch)
 	}
 
-	if vmStats.Memory.SwapInSet || vmStats.Memory.SwapOutSet {
-		swapTrafficLabels := []string{"node", "namespace", "name", "type"}
-		swapTrafficLabels = append(swapTrafficLabels, metrics.k8sLabels...)
-		swapTrafficDesc := prometheus.NewDesc(
+	if mem.SwapInSet || mem.SwapOutSet {
+		desc := metrics.newPrometheusDesc(
 			"kubevirt_vmi_memory_swap_traffic_bytes_total",
 			"swap memory traffic.",
-			swapTrafficLabels,
-			nil,
+			[]string{"type"},
 		)
 
-		if vmStats.Memory.SwapInSet {
-			swapTrafficInLabelValues := []string{metrics.vmi.Status.NodeName, metrics.vmi.Namespace, metrics.vmi.Name, "in"}
-			swapTrafficInLabelValues = append(swapTrafficInLabelValues, metrics.k8sLabelValues...)
-
-			mv, err := prometheus.NewConstMetric(
-				swapTrafficDesc, prometheus.GaugeValue,
-				// the libvirt value is in KiB
-				float64(vmStats.Memory.SwapIn)*1024,
-				swapTrafficInLabelValues...,
-			)
-			tryToPushMetric(swapTrafficDesc, mv, err, metrics.ch)
+		if mem.SwapInSet {
+			metrics.pushPrometheusMetric(desc, prometheus.GaugeValue, float64(mem.SwapIn)*1024, []string{"in"})
 		}
-		if vmStats.Memory.SwapOutSet {
-			swapTrafficOutLabelValues := []string{metrics.vmi.Status.NodeName, metrics.vmi.Namespace, metrics.vmi.Name, "out"}
-			swapTrafficOutLabelValues = append(swapTrafficOutLabelValues, metrics.k8sLabelValues...)
-
-			mv, err := prometheus.NewConstMetric(
-				swapTrafficDesc, prometheus.GaugeValue,
-				// the libvirt value is in KiB
-				float64(vmStats.Memory.SwapOut)*1024,
-				swapTrafficOutLabelValues...,
-			)
-			tryToPushMetric(swapTrafficDesc, mv, err, metrics.ch)
+		if mem.SwapOutSet {
+			metrics.pushPrometheusMetric(desc, prometheus.GaugeValue, float64(mem.SwapOut)*1024, []string{"out"})
 		}
 	}
 }
 
-func (metrics *vmiMetrics) updateVcpu(vmStats *stats.DomainStats) {
-	for vcpuId, vcpu := range vmStats.Vcpu {
-		// Initial vcpu metrics labels
-		if !vcpu.StateSet || !vcpu.TimeSet {
-			log.Log.V(4).Warningf("State or time not set for vcpu#%d", vcpuId)
```

Comment:
do we still want to have this warning log printed?

Reply:
Honestly, I don't think we need it @ashleyschuett - we don't print this warning in any other metric